File: enthropic/state.py
from __future__ import annotations

from pathlib import Path

from .parser import EnthSpec

STATUS_VALUES = {"BUILT", "PARTIAL", "PENDING", "OK", "MISSING", "UNVERIFIED", "SET", "UNSET"}


def generate(spec: EnthSpec, project_name: str) -> str:
    """Generate a structured state file from a spec. All items start as PENDING/UNVERIFIED."""
    lines = [f"STATE {project_name}", ""]

    # CHECKS — derived from LANG + DEPS.SYSTEM + DEPS.RUNTIME
    checks: list[tuple[str, str]] = []
    if spec.project.get("LANG"):
        checks.append((spec.project["LANG"], "LANG"))
    deps = spec.project.get("DEPS", {})
    for dep in deps.get("SYSTEM", []):
        checks.append((dep, "DEPS.SYSTEM"))
    for dep in deps.get("RUNTIME", []):
        checks.append((dep, "DEPS.RUNTIME"))

    if checks:
        lines.append("  CHECKS")
        for name, source in checks:
            lines.append(f"    {name:<28} UNVERIFIED   # {source}")
        lines.append("")

    if spec.entities:
        lines.append("  ENTITY")
        for entity in spec.entities:
            lines.append(f"    {entity:<28} PENDING")
        lines.append("")

    if spec.flows:
        lines.append("  FLOWS")
        for name in spec.flows:
            lines.append(f"    {name:<28} PENDING")
        lines.append("")

    if spec.layers:
        lines.append("  LAYERS")
        for name in spec.layers:
            lines.append(f"    {name:<28} PENDING")
        lines.append("")

    return "\n".join(lines)


def load(path: Path) -> dict:
    """Parse a state file into a dict of section -> {key: status}."""
    state: dict[str, dict[str, str]] = {"entity": {}, "flows": {}, "layers": {}}
    section: str | None = None

    for line in path.read_text(encoding="utf-8").splitlines():
        tok = line.strip()
        if not tok or tok.startswith("STATE "):
            continue
        if tok in ("ENTITY", "FLOWS", "LAYERS"):
            section = tok.lower()
            continue
        if section:
            parts = tok.split()
            if len(parts) == 2 and parts[1] in STATUS_VALUES:
                state[section][parts[0]] = parts[1]

    return state


def set_status(path: Path, key: str, status: str, section: str = "") -> None:
    """Update a single entry in the state file.

    key: the entity/flow/layer name
    status: BUILT | PARTIAL | PENDING
    """
    status = status.upper()
    if status not in STATUS_VALUES:
        raise ValueError(f"Invalid status '{status}'. Must be: {', '.join(sorted(STATUS_VALUES))}")

    lines = path.read_text(encoding="utf-8").splitlines()
    result = []
    updated = False

    for line in lines:
        tok = line.strip()
        parts = tok.split()
        if len(parts) == 2 and parts[0] == key and parts[1] in STATUS_VALUES:
            indent = line[: len(line) - len(line.lstrip())]
            result.append(f"{indent}{key:<28} {status}")
            updated = True
        else:
            result.append(line)

    if not updated:
        raise KeyError(f"Key '{key}' not found in state file")

    path.write_text("\n".join(result) + "\n", encoding="utf-8")

File: src/state.rs
use std::path::Path;
use anyhow::Result;
use crate::parser::{EnthSpec, ProjectValue};

pub const STATUS_VALUES: &[&str] = &[
    "BUILT", "PARTIAL", "PENDING", "OK", "MISSING", "UNVERIFIED", "SET", "UNSET",
];

pub fn generate(spec: &EnthSpec, project_name: &str) -> String {
    let mut lines: Vec<String> = vec![format!("STATE {}", project_name), String::new()];

    // CHECKS — derived from LANG + DEPS.SYSTEM + DEPS.RUNTIME
    let mut checks: Vec<(String, String)> = Vec::new();
    if let Some(ProjectValue::Str(lang)) = spec.project.get("LANG") {
        checks.push((lang.clone(), "LANG".to_string()));
    }
    if let Some(ProjectValue::Deps(deps)) = spec.project.get("DEPS") {
        for dep in deps.get("SYSTEM").unwrap_or(&vec![]) {
            checks.push((dep.clone(), "DEPS.SYSTEM".to_string()));
        }
        for dep in deps.get("RUNTIME").unwrap_or(&vec![]) {
            checks.push((dep.clone(), "DEPS.RUNTIME".to_string()));
        }
    }

    if !checks.is_empty() {
        lines.push("  CHECKS".to_string());
        for (name, source) in &checks {
            lines.push(format!("    {:<28} UNVERIFIED   # {}", name, source));
        }
        lines.push(String::new());
    }

    if !spec.entities.is_empty() {
        lines.push("  ENTITY".to_string());
        for entity in &spec.entities {
            lines.push(format!("    {:<28} PENDING", entity));
        }
        lines.push(String::new());
    }

    if !spec.flows.is_empty() {
        lines.push("  FLOWS".to_string());
        for name in &spec.flows_order {
            lines.push(format!("    {:<28} PENDING", name));
        }
        lines.push(String::new());
    }

    if !spec.layers.is_empty() {
        lines.push("  LAYERS".to_string());
        for name in &spec.layers_order {
            lines.push(format!("    {:<28} PENDING", name));
        }
        lines.push(String::new());
    }

    lines.join("\n")
}

pub fn set_status(path: &Path, key: &str, status: &str) -> Result<()> {
    let status = status.to_uppercase();
    if !STATUS_VALUES.contains(&status.as_str()) {
        anyhow::bail!(
            "Invalid status '{}'. Must be: {}",
            status,
            STATUS_VALUES.join(", ")
        );
    }

    let content = std::fs::read_to_string(path)?;
    let mut result: Vec<String> = Vec::new();
    let mut updated = false;

    for line in content.lines() {
        let tok = line.trim();
        let parts: Vec<&str> = tok.split_whitespace().collect();
        if parts.len() == 2 && parts[0] == key && STATUS_VALUES.contains(&parts[1]) {
            let leading_len = line.len() - line.trim_start().len();
            let leading = &line[..leading_len];
            result.push(format!("{}{:<28} {}", leading, key, status));
            updated = true;
        } else {
            result.push(line.to_string());
        }
    }

    if !updated {
        anyhow::bail!("Key '{}' not found in state file", key);
    }

    std::fs::write(path, result.join("\n") + "\n")?;
    Ok(())
}
